### PR TITLE
Upgrade stylish-haskell to 0.12.2.2

### PR DESCRIPTION
**Pull request by EncodePanda**
Fixes https://github.com/input-output-hk/ouroboros-network/issues/2951

Some of the important changes in stylish since 0.11.0.0

*  Use ghc-lib-parser rather than haskell-src-exts (will work with 8.10.x features)
* when wrapping import lists over multiple lines, you can repeat the module name
* a new Step was added, `module_header`, which formats the export list of a module
* a new option for aligning only adjacent items
* support for aligning MultiWayIf syntax
  
Opens up new possibilities like https://github.com/input-output-hk/ouroboros-network/issues/2952

```
$> nix-shell

$> stylish-haskell --version
stylish-haskell 0.12.2.0
```

### CODE REVIEW

File: nix/stylish-haskell.nix [hunk 1]
```diff
@@ -0,0 +1,16 @@
+{ pkgs
```
Comment (nfrisby) [inline nix/stylish-haskell.nix:1]: The current `stylish-haskell` is pulled in here

https://github.com/input-output-hk/ouroboros-network/blob/e142c7153b9ff0505a84ec68124f5bd80449eebd/shell.nix#L30

May we can make the change on that line? If not, we'd at least have to remove that, probably.

I don't know the details of this. Marcin might, or else the devops-crossteam-ci Slack channel eg.

Comment (nfrisby) [reply nix/stylish-haskell.nix:1]: Ah, I see now that you removed it below. I suppose setting that to `0.12.2.0` didn't work?

I wonder which part of `haskell.nix` upstream would have to update in order to support just changing that line.

Comment (EncodePanda) [reply nix/stylish-haskell.nix:1]: modifying it to 0.12.2.0 did not work, those are some "hard-coded" tools within haskell.nix.
it's just easier to point to an existing release

Comment (EncodePanda) [reply nix/stylish-haskell.nix:1]: Oh, and one more thing, with some things I'm working on atm in stylish we will most likely for a while work on an "unreleased" `0.12.3.0`

File: ouroboros-consensus-cardano/src/Ouroboros/Consensus/Cardano/CanHardFork.hs [hunk 2]
```diff
@@ -31,7 +31,7 @@ import           Control.Monad.Except (Except, runExcept, throwError)
 import qualified Data.Map.Strict as Map
 import           Data.Maybe (listToMaybe, mapMaybe)
 import           Data.Proxy
-import           Data.SOP.Strict ((:.:) (..), NP (..), unComp)
```
Comment (nfrisby) [inline ouroboros-consensus-cardano/src/Ouroboros/Consensus/Cardano/CanHardFork.hs:34]: Punctuation vs alphabet seemed to have changed. (Not list in stylish-haskell CHANGELOG.)

File: ouroboros-consensus-test/src/Test/ThreadNet/Network.hs [hunk 3]
```diff
@@ -410,7 +410,7 @@ runThreadNetwork systemTime ThreadNetworkArgs
       forM vertexInfos0 $ \(coreNodeId, vertexStatusVar, readNodeInfo) -> do
         readTVar vertexStatusVar >>= \case
           VDown ch ldgr -> pure (coreNodeId, readNodeInfo, ch, ldgr)
-          _        -> retry
+          _             -> retry
```
Comment (nfrisby) [inline ouroboros-consensus-test/src/Test/ThreadNet/Network.hs:413]: I don't have any idea where this whitespace is coming from.

File: ouroboros-consensus/src/Ouroboros/Consensus/Block/Abstract.hs [hunk 4]
```diff
@@ -74,11 +74,11 @@ import           Cardano.Slotting.Slot (EpochNo (..), EpochSize (..),
...
 
-import           Ouroboros.Network.Block (pattern BlockPoint, ChainHash (..),
-                     pattern GenesisPoint, HasHeader (..), HeaderFields (..),
-                     HeaderHash, Point, StandardHash, blockHash, blockNo,
-                     blockPoint, blockSlot, castHash, castHeaderFields,
-                     castPoint, pointHash, pointSlot)
+import           Ouroboros.Network.Block (ChainHash (..), HasHeader (..),
+                     HeaderFields (..), HeaderHash, Point, StandardHash,
+                     blockHash, blockNo, blockPoint, blockSlot, castHash,
+                     castHeaderFields, castPoint, pattern BlockPoint,
```
Comment (nfrisby) [inline ouroboros-consensus/src/Ouroboros/Consensus/Block/Abstract.hs:80]: It's not clear to me what happened here with the pattern synonyms.

File: ouroboros-consensus-byron-test/src/Ouroboros/Consensus/ByronDual/Node.hs [hunk 5]
```diff
@@ -18,8 +18,8 @@ import           Data.Maybe (fromMaybe)
 
 import qualified Byron.Spec.Ledger.Core as Spec
 import qualified Byron.Spec.Ledger.Delegation as Spec
-import qualified Byron.Spec.Ledger.Update as Spec
 import qualified Byron.Spec.Ledger.UTxO as Spec
+import qualified Byron.Spec.Ledger.Update as Spec
```
Comment (nfrisby) [inline ouroboros-consensus-byron-test/src/Ouroboros/Consensus/ByronDual/Node.hs:23]: Looks like the old comparison ignored case. (Not list in stylish-haskell CHANGELOG.)

File: nix/stylish-haskell.nix (hunk 1 above)
Comment (nfrisby) [reply nix/stylish-haskell.nix:1]: Looks like `tools` should pick up things that are on Hackage.

https://github.com/input-output-hk/haskell.nix/blob/3f9e7d746895aa9811867a85a3790ba1e3da2798/overlays/tools.nix#L96-L105

Jasper uploaded `0.12.2.0` for the 2020-10-08T12:49:46Z hackage timestamp. The `cabal` file on this branch specifies `index-state: 2020-12-11T00:00:00Z`, which is later. I wonder why that isn't enough for `tools` to find that `stylish-haskell`. :shrug:

Maybe that's a `tools` bug?

Comment (nfrisby) [reply nix/stylish-haskell.nix:1]: Actually, looks like it works on `master` as of today (maybe the index-state was increased recently? /shrug).

```
nfrisby@frisbycomp:~/ouroboros-network$ git lg -1
59e71cb54 - (HEAD, origin/master, origin/bors/staging, origin/HEAD) Merge #2965 (4 days ago) <iohk-bors[bot]>
nfrisby@frisbycomp:~/ouroboros-network$ git diff
diff --git a/shell.nix b/shell.nix
index b25df69a9..b70e01dd8 100644
--- a/shell.nix
+++ b/shell.nix
@@ -27,7 +27,7 @@ let
     ];
 
     tools = {
-      stylish-haskell = "0.11.0.0";
+      stylish-haskell = "0.12.2.0";
       ghcid = "0.8.7";
       cabal = "3.2.0.0";
       # todo: add back the build tools which are actually necessary
nfrisby@frisbycomp:~/ouroboros-network$ nix-shell --run 'stylish-haskell --version' 2>&1 | tail -n1
stylish-haskell 0.12.2.0
```

**Comment by EncodePanda** [comment]
Regarding imports sorting that changed between 0.11.x.x and 0.12.x.x

 `lib/Language/Haskell/Stylish/Ordering.hs` is responsible for sorting elements within the single import list and
   is providing its own `Ord` so will most likely not change in the future (stylish controls the `Ord` instance).
   
   However the list of imports is sorted by ~Ord~ [provided by ghc-parser ](https://github.com/jaspervdj/stylish-haskell/blob/a996292f56810f96372c012ccef6adb52cdab487/lib/Language/Haskell/Stylish/Step/Imports.hs#L443). There will be a ticket in stylish-haskell to unify this so that stylish has control over how imports are sorted (and not delegate that to a dependency).

**Comment by EncodePanda** [comment]
@nfrisby I've managed to unify the GHC version, so it is defined in a single place

**Comment by EncodePanda** [comment]
Closing this PR. We've upgraded stylish haskell just by upgrading version number in 'tools' in https://github.com/input-output-hk/ouroboros-network/pull/2981.

Nevertheless two things came out from the work on this ticket:
1. A need to have a nix formatter (see https://github.com/input-output-hk/ouroboros-network/issues/2980)
2. A single place to define GHC version in Nix (see https://github.com/input-output-hk/ouroboros-network/issues/2982)

**Pull request closed without merging**